File: cordova-plugin-casai/src/android/CASMobileAds.kt
package com.cleveradssolutions.plugin.cordova

import android.app.Activity
import android.content.res.Configuration
import com.cleveradssolutions.sdk.AdContentInfo
import com.cleveradssolutions.sdk.AdFormat
import com.cleveradssolutions.sdk.AdRevenuePrecision
import com.cleveradssolutions.sdk.base.CASHandler
import com.cleveradssolutions.sdk.screen.CASAppOpen
import com.cleveradssolutions.sdk.screen.CASInterstitial
import com.cleveradssolutions.sdk.screen.CASRewarded
import com.cleversolutions.ads.AdError
import com.cleversolutions.ads.AdSize
import com.cleversolutions.ads.Audience
import com.cleversolutions.ads.ConsentFlow
import com.cleversolutions.ads.TargetingOptions
import com.cleversolutions.ads.android.CAS
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaPlugin
import org.json.JSONArray
import org.json.JSONObject


internal object PluginEvents {
    const val LOADED = "casai_ad_loaded"
    const val LOAD_FAILED = "casai_ad_load_failed"
    const val SHOWED = "casai_ad_showed"
    const val SHOW_FAILED = "casai_ad_show_failed"
    const val CLICKED = "casai_ad_clicked"
    const val IMPRESSIONS = "casai_ad_impressions"
    const val DISMISSED = "casai_ad_dismissed"
    const val REWARD = "casai_ad_reward"
}

class CASMobileAds : CordovaPlugin() {
    var casId: String = ""
        private set

    val activity: Activity? get() = cordova.activity

    private val banner = ViewAdManager(this, AdFormat.BANNER)
    private val mrec = ViewAdManager(this, AdFormat.MEDIUM_RECTANGLE)

    private var interstitial = ScreenAdManager(this, AdFormat.INTERSTITIAL)
    private var rewarded = ScreenAdManager(this, AdFormat.REWARDED)
    private var appOpen = ScreenAdManager(this, AdFormat.APP_OPEN)

    private var pendingInitCallback: CallbackContext? = null
    private var initResult: JSONObject? = null

    override fun pluginInitialize() {
        super.pluginInitialize()
        casId = cordova.context.applicationContext.packageName
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        banner.onConfigurationChanged(newConfig)
        mrec.onConfigurationChanged(newConfig)
    }

    override fun execute(
        action: String,
        data: JSONArray,
        callbackContext: CallbackContext
    ): Boolean {
        when (action) {
            "initialize" -> initialize(data, callbackContext)
            "showConsentFlow" -> showConsentFlow(data, callbackContext)
            "getSDKVersion" -> callbackContext.success(CAS.getSDKVersion())
            "setDebugLoggingEnabled" -> {
                CAS.settings.debugMode = data.optBoolean(0, false)
                callbackContext.success()
            }

            "setAdSoundsMuted" -> {
                CAS.settings.mutedAdSounds = data.optBoolean(0, false)
                callbackContext.success()
            }

            "setUserAge" -> {
                CAS.targetingOptions.age = data.optInt(0, 0)
                callbackContext.success()
            }

            "setUserGender" -> {
                CAS.targetingOptions.gender = when (data.optString(0, null)) {
                    "male" -> TargetingOptions.GENDER_MALE
                    "female" -> TargetingOptions.GENDER_FEMALE
                    else -> TargetingOptions.GENDER_UNKNOWN
                }
                callbackContext.success()
            }

            "setAppKeywords" -> {
                val arr = data.optJSONArray(0) ?: JSONArray()
                CAS.targetingOptions.keywords =
                    (0 until arr.length()).map { arr.optString(it) }.toSet()
                callbackContext.success()
            }

            "setAppContentUrl" -> {
                CAS.targetingOptions.contentUrl =
                    if (data.isNull(0)) null else data.optString(0, null)
                callbackContext.success()
            }

            "setLocationCollectionEnabled" -> {
                CAS.targetingOptions.locationCollectionEnabled =
                    data.optBoolean(0, false);
                callbackContext.success()
            }

            "setTrialAdFreeInterval" -> {
                CAS.settings.trialAdFreeInterval = data.optInt(0, 0);
                callbackContext.success()
            }

            "loadBannerAd" -> loadBannerAd(data, callbackContext)
            "showBannerAd" -> banner.show(data, callbackContext)
            "hideBannerAd" -> banner.hide(callbackContext)
            "destroyBannerAd" -> banner.destroy(callbackContext)

            "loadMRecAd" -> loadMRecAd(data, callbackContext)
            "showMRecAd" -> mrec.show(data, callbackContext)
            "hideMRecAd" -> mrec.hide(callbackContext)
            "destroyMRecAd" -> mrec.destroy(callbackContext)

            "loadAppOpenAd" -> loadAppOpenAd(data, callbackContext)
            "isAppOpenAdLoaded" -> appOpen.sendIsLoaded(callbackContext)
            "showAppOpenAd" -> showAppOpenAd(callbackContext)
            "destroyAppOpenAd" -> appOpen.destroyAd(callbackContext)

            "loadInterstitialAd" -> loadInterstitialAd(data, callbackContext)
            "isInterstitialAdLoaded" -> interstitial.sendIsLoaded(callbackContext)
            "showInterstitialAd" -> showInterstitialAd(callbackContext)
            "destroyInterstitialAd" -> interstitial.destroyAd(callbackContext)

            "loadRewardedAd" -> loadRewardedAd(data, callbackContext)
            "isRewardedAdLoaded" -> rewarded.sendIsLoaded(callbackContext)
            "showRewardedAd" -> showRewardedAd(callbackContext)
            "destroyRewardedAd" -> rewarded.destroyAd(callbackContext)

            else -> return false
        }
        return true
    }

    private fun debugGeoFrom(s: String): Int = when (s.lowercase()) {
        "eea" -> ConsentFlow.DebugGeography.EEA
        "us" -> ConsentFlow.DebugGeography.REGULATED_US_STATE
        "unregulated", "other" -> ConsentFlow.DebugGeography.OTHER
        else -> ConsentFlow.DebugGeography.DISABLED
    }

    private fun consentStatusText(@ConsentFlow.Status s: Int): String = when (s) {
        ConsentFlow.Status.UNKNOWN -> "Unknown"
        ConsentFlow.Status.OBTAINED -> "Obtained"
        ConsentFlow.Status.NOT_REQUIRED -> "Not required"
        ConsentFlow.Status.UNAVAILABLE -> "Unavailable"
        ConsentFlow.Status.INTERNAL_ERROR -> "Internal error"
        ConsentFlow.Status.NETWORK_ERROR -> "Network error"
        ConsentFlow.Status.CONTEXT_INVALID -> "Invalid context"
        ConsentFlow.Status.FLOW_STILL_SHOWING -> "Still presenting"
        else -> "Unknown"
    }

    private fun initialize(args: JSONArray, callbackContext: CallbackContext) {
        initResult?.let {
            callbackContext.success(it)
            return
        }

        pendingInitCallback = callbackContext

        val cordovaVersion = args.getString(0)
        val frameworkName = args.getString(1)
        val targetAudience = args.optString(2)
        val showConsentForm = args.getBoolean(3)
        val forceTestAds = args.getBoolean(4)
        val testDeviceIds = args.optJSONArray(5)
        val debugGeography = args.optString(6)
        val mediationExtras = args.optJSONObject(7)

        when (targetAudience) {
            "children" -> CAS.settings.taggedAudience = Audience.CHILDREN
            "notchildren" -> CAS.settings.taggedAudience = Audience.NOT_CHILDREN
        }

        if (testDeviceIds != null && testDeviceIds.length() > 0) {
            val testDevices = HashSet<String>()
            for (index in 0 until testDeviceIds.length()) {
                testDevices.add(testDeviceIds.optString(index))
            }
            CAS.settings.testDeviceIDs = testDevices
        }

        val consentFlow = ConsentFlow(showConsentForm)
            .withUIContext(activity)
            .withForceTesting(forceTestAds)
        if (debugGeography.isNotEmpty()) {
            consentFlow.withDebugGeography(debugGeoFrom(debugGeography))
        }

        val managerBuilder = CAS.buildManager()
            .withCasId(casId)
            .withTestAdMode(forceTestAds)
            .withFramework(frameworkName, cordovaVersion)
            .withConsentFlow(consentFlow)
            .withCompletionListener { configuration ->
                initResult = JSONObject().apply {
                    configuration.error?.let { put("error", it) }
                    configuration.countryCode?.let { put("countryCode", it) }
                    put("isConsentRequired", configuration.isConsentRequired)
                    put("consentFlowStatus", consentStatusText(configuration.consentFlowStatus))
                }

                pendingInitCallback?.let { cb ->
                    pendingInitCallback = null
                    cb.success(initResult)
                }
            }

        mediationExtras?.keys()?.forEach { key ->
            managerBuilder.withMediationExtras(key, mediationExtras.optString(key))
        }
        managerBuilder.build(cordova.activity ?: cordova.context)
    }

    private fun showConsentFlow(args: JSONArray, callbackContext: CallbackContext) {
        val ifRequired = args.optBoolean(0, false)
        val debugGeography = args.optString(1, null)

        val flow = ConsentFlow(ifRequired)
            .withUIContext(activity)
            .withDismissListener { status ->
                callbackContext.success(consentStatusText(status))
            }
        if (debugGeography != null) {
            val type = debugGeoFrom(debugGeography)
            if (type != ConsentFlow.DebugGeography.DISABLED) {
                flow.withDebugGeography(type)
            }
        }

        if (ifRequired) flow.showIfRequired() else flow.show()
    }

    private fun loadBannerAd(args: JSONArray, callbackContext: CallbackContext) {
        val sizeCode = args.getString(0)
        val maxWidthDp = args.getInt(1)
        val maxHeightDp = args.getInt(2)
        val autoload = args.getBoolean(3)
        val refreshInterval = args.getInt(4)

        val adSize = banner.resolveAdSize(
            sizeCode[0], maxWidthDp, maxHeightDp
        )
        banner.loadBanner(
            adSize, autoload, refreshInterval, callbackContext
        )
    }

    private fun loadMRecAd(args: JSONArray, cb: CallbackContext) {
        val autoload = args.getBoolean(0)
        val refresh = args.getInt(1)
        mrec.loadBanner(AdSize.MEDIUM_RECTANGLE, autoload, refresh, cb)
    }

    private fun loadAppOpenAd(args: JSONArray, callbackContext: CallbackContext) {
        val autoload = args.getBoolean(0)
        val autoShow = args.getBoolean(1)
        val ad = appOpen.ad as? CASAppOpen ?: CASAppOpen(casId)
        ad.isAutoshowEnabled = autoShow
        appOpen.loadAd(ad, autoload, callbackContext)
    }

    private fun showAppOpenAd(callbackContext: CallbackContext) {
        val ad = appOpen.beforeShowAd(callbackContext) as? CASAppOpen
        ad?.show(activity)
    }

    private fun loadInterstitialAd(args: JSONArray, callbackContext: CallbackContext) {
        val autoload = args.getBoolean(0)
        val autoShow = args.getBoolean(1)
        val minIntervalSec = args.getInt(2)

        val ad = interstitial.ad as? CASInterstitial ?: CASInterstitial(casId)
        ad.isAutoshowEnabled = autoShow
        ad.minInterval = minIntervalSec
        interstitial.loadAd(ad, autoload, callbackContext)
    }

    private fun showInterstitialAd(callbackContext: CallbackContext) {
        val ad = interstitial.beforeShowAd(callbackContext) as? CASInterstitial
        ad?.show(activity)
    }

    private fun loadRewardedAd(args: JSONArray, callbackContext: CallbackContext) {
        val autoload = args.getBoolean(0)

        val ad = rewarded.ad as? CASRewarded ?: CASRewarded(casId)
        rewarded.loadAd(ad, autoload, callbackContext)
    }

    private fun showRewardedAd(callbackContext: CallbackContext) {
        val ad = rewarded.beforeShowAd(callbackContext) as? CASRewarded
        ad?.show(activity, rewarded)
    }

    fun emitErrorEvent(type: String, format: AdFormat, error: AdError, callback: CallbackContext?) {
        val json = errorJson(format, error)
        emitEvent(type, json)
        callback?.error(json)
    }

    fun emitEvent(type: String, format: AdFormat) {
        emitEvent(type, JSONObject().put("format", format.label))
    }

    fun emitEvent(type: String, payload: JSONObject) {
        val js = "cordova.fireDocumentEvent(${JSONObject.quote(type)}, $payload);"
        CASHandler.main {
            webView.engine?.evaluateJavascript(js, null)
                ?: webView.loadUrl("javascript:$js")
        }
    }

    fun emitImpressionEvent(format: AdFormat, info: AdContentInfo) {
        val precision = when (info.revenuePrecision) {
            AdRevenuePrecision.PRECISE -> "precise"
            AdRevenuePrecision.FLOOR -> "floor"
            AdRevenuePrecision.ESTIMATED -> "estimated"
            else -> "unknown"
        }

        val json = JSONObject()
            .put("format", format.label)
            .put("sourceUnitId", info.sourceUnitId)
            .put("sourceName", info.sourceName)
            .put("revenue", info.revenue)
            .put("revenuePrecision", precision)
            .put("revenueTotal", info.revenueTotal)
            .put("impressionDepth", info.impressionDepth)
        info.creativeId?.let {
            json.put("creativeId", it)
        }
        emitEvent(PluginEvents.IMPRESSIONS, json)
    }

    fun errorJson(format: AdFormat, error: AdError): JSONObject =
        JSONObject()
            .put("format", format.label)
            .put("code", error.code)
            .put("message", error.message)

    fun cancelledLoadError(format: AdFormat): JSONObject =
        errorJson(format, AdError(499, "Load Promise interrupted by new load call"))
}

File: cordova-plugin-casai/src/ios/CASMobileAds.swift
import CleverAdsSolutions
import Foundation

enum CASEvent: String {
    case casai_ad_loaded
    case casai_ad_showed

    case casai_ad_load_failed
    case casai_ad_show_failed

    case casai_ad_impressions

    case casai_ad_clicked
    case casai_ad_dismissed

    case casai_ad_reward
}

@objc(CASMobileAds)
class CASMobileAds: CDVPlugin {
    private(set) var casId: String = ""
    private var initResponse: [String: Any] = [:]

    private var mrecManager: CASViewAdManager
    private var bannerManager: CASViewAdManager

    private var interstitialManager: CASScreenAdManager
    private var rewardedManager: CASScreenAdManager
    private var appOpenManager: CASScreenAdManager

    override init() {
        mrecManager = CASViewAdManager(format: .mediumRectangle)
        bannerManager = CASViewAdManager(format: .banner)
        interstitialManager = CASScreenAdManager(format: .interstitial)
        rewardedManager = CASScreenAdManager(format: .rewarded)
        appOpenManager = CASScreenAdManager(format: .appOpen)
    }

    /// Called after plugin construction and fields have been initialized.
    override func pluginInitialize() {
        super.pluginInitialize()

        if let id = Bundle.main.object(
            forInfoDictionaryKey: "CASAIAppIdentifier"
        ) as? String {
            casId = id
        }

        mrecManager.plugin = self
        bannerManager.plugin = self
        interstitialManager.plugin = self
        rewardedManager.plugin = self
        appOpenManager.plugin = self
    }

    // MARK: - Init

    @objc func initialize(_ command: CDVInvokedUrlCommand) {
        // nativePromise('initialize', [
        // /* 0 */ cordova.version,
        // /* 1 */ overrideFramework,
        // /* 2 */ targetAudience,
        // /* 3 */ showConsentFormIfRequired ?? true,
        // /* 4 */ forceTestAds ?? false,
        // /* 5 */ testDeviceIds,
        // /* 6 */ debugGeography ?? 'eea',
        // /* 7 */ mediationExtras,

        var callbackId: String? = command.callbackId

        let cordovaVersion = command.arguments[0] as! String
        let frameworkName = command.arguments[1] as! String
        let targetAudience = command.arguments[2] as? String
        let showConsentForm = command.arguments[3] as! Bool
        let forceTestAds = command.arguments[4] as! Bool
        let testDeviceIds = command.arguments[5] as? [String]
        let debugGeography = command.arguments[6] as? String
        let mediationExtras = command.arguments[7] as? [String: Any]

        if !initResponse.isEmpty {
            let result = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: initResponse
            )
            commandDelegate?.send(result, callbackId: callbackId)
            return
        }

        let consentFlow = CASConsentFlow()
        consentFlow.isEnabled = showConsentForm
        consentFlow.forceTesting = forceTestAds

        if let testDeviceIds, !testDeviceIds.isEmpty {
            CAS.settings.setTestDevice(ids: testDeviceIds)
        }

        if let targetAudience {
            switch targetAudience.lowercased() {
            case "children":
                CAS.settings.taggedAudience = .children
            case "notchildren":
                CAS.settings.taggedAudience = .notChildren
            default: break
            }
        }

        switch debugGeography?.lowercased() {
        case "eea":
            consentFlow.debugGeography = .EEA
        case "us":
            consentFlow.debugGeography = .regulatedUSState
        case "unregulated", "other":
            consentFlow.debugGeography = .other
        default: break
        }

        let builder = CAS.buildManager()
            .withTestAdMode(forceTestAds)
            .withConsentFlow(consentFlow)
            .withFramework(frameworkName, version: cordovaVersion)

        if let mediationExtras {
            for (key, value) in mediationExtras {
                if let strValue = value as? String {
                    builder.withMediationExtras(strValue, forKey: key)
                }
            }
        }

        builder.withCompletionHandler { config in
            self.initResponse["error"] = config.error
            self.initResponse["countryCode"] = config.countryCode
            self.initResponse["isConsentRequired"] = config.isConsentRequired
            self.initResponse["consentFlowStatus"] = self.getConsentFlowStatus(
                from: config.consentFlowStatus
            )

            if let id = callbackId {
                let result = CDVPluginResult(
                    status: CDVCommandStatus_OK,
                    messageAs: self.initResponse
                )
                self.commandDelegate?.send(result, callbackId: id)
                callbackId = nil
            }
        }
        .create(withCasId: casId)
    }

    // MARK: - ConsentFlow

    @objc func showConsentFlow(_ command: CDVInvokedUrlCommand) {
        // nativePromise('showConsentFlow', [ifRequired, debugGeography]);
        let ifRequired = command.arguments[0] as? Bool ?? false
        let debugGeography = command.arguments[1] as? String

        let consentFlow = CASConsentFlow(isEnabled: true)
            .withViewControllerToPresent(viewController)
            .withCompletionHandler { status in
                self.sendOk(
                    command.callbackId,
                    messageAs: self.getConsentFlowStatus(from: status)
                )
            }

        switch debugGeography?.lowercased() {
        case "eea":
            consentFlow.debugGeography = .EEA
        case "us":
            consentFlow.debugGeography = .regulatedUSState
        case "unregulated", "other":
            consentFlow.debugGeography = .other
        default: break
        }

        if ifRequired {
            consentFlow.presentIfRequired()
        } else {
            consentFlow.present()
        }
    }

    // MARK: - Additional Methods

    @objc func getSDKVersion(_ command: CDVInvokedUrlCommand) {
        // nativePromise('getSDKVersion');
        sendOk(command.callbackId, messageAs: CAS.getSDKVersion())
    }

    @objc func setDebugLoggingEnabled(_ command: CDVInvokedUrlCommand) {
        // nativeCall('setDebugLoggingEnabled', [enabled]);
        CAS.settings.debugMode = command.argument(at: 0) as! Bool
        sendOk(command.callbackId)
    }

    @objc func setAdSoundsMuted(_ command: CDVInvokedUrlCommand) {
        // nativeCall('setAdSoundsMuted', [muted]);
        CAS.settings.mutedAdSounds = command.argument(at: 0) as! Bool
        sendOk(command.callbackId)
    }

    @objc func setUserAge(_ command: CDVInvokedUrlCommand) {
        // nativeCall('setUserAge', [age]);
        CAS.targetingOptions.age = command.argument(at: 0) as! Int
        sendOk(command.callbackId)
    }

    @objc func setUserGender(_ command: CDVInvokedUrlCommand) {
        // nativeCall('setUserGender', [gender]);
        let gender = command.argument(at: 0) as? String
        if gender == "male" {
            CAS.targetingOptions.gender = .male
        } else if gender == "female" {
            CAS.targetingOptions.gender = .female
        } else {
            CAS.targetingOptions.gender = .unknown
        }
        sendOk(command.callbackId)
    }

    @objc func setAppKeywords(_ command: CDVInvokedUrlCommand) {
        // nativeCall('setAppKeywords', [keywords]);
        if let keywords = command.argument(at: 0) as? [String] {
            CAS.targetingOptions.keywords = keywords
        }
        sendOk(command.callbackId)
    }

    @objc func setAppContentUrl(_ command: CDVInvokedUrlCommand) {
        // nativeCall('setAppContentUrl', [contentUrl]);
        CAS.targetingOptions.contentUrl = command.argument(at: 0) as? String
        sendOk(command.callbackId)
    }

    @objc func setLocationCollectionEnabled(_ command: CDVInvokedUrlCommand) {
        // nativeCall('setLocationCollectionEnabled', [enabled]);
        CAS.targetingOptions.locationCollectionEnabled = command.argument(at: 0) as! Bool
        sendOk(command.callbackId)
    }

    @objc func setTrialAdFreeInterval(_ command: CDVInvokedUrlCommand) {
        // nativeCall('setTrialAdFreeInterval', [interval]);
        CAS.settings.trialAdFreeInterval =
            (command.argument(at: 0) as! NSNumber).uint64Value
        sendOk(command.callbackId)
    }

    // MARK: - Banner Ad

    /// Init and Load banner. command.arguments should be:
    /// [ adSizeString: String, maxWidth: Double?, maxHeight: Double?, autoReload: Bool?, refreshInterval: Int? ]
    @objc func loadBannerAd(_ command: CDVInvokedUrlCommand) {
        // nativePromise('loadBannerAd', [
        // adSize ?? 'S',
        // maxWidth ?? 0,
        // maxHeight ?? 0,
        // autoReload ?? true,
        // refreshInterval ?? 30,
        // ]);
        let adSizeString = command.arguments[0] as! String
        let maxWidth = command.arguments[1] as! NSNumber
        let maxHeight = command.arguments[2] as! NSNumber
        let autoReload = command.arguments[3] as! Bool
        let refreshInterval = command.arguments[4] as! NSNumber

        bannerManager.loadAd(
            bannerManager.resolveAdSize(
                adSizeString,
                maxWidth: maxWidth.intValue,
                maxHeight: maxHeight.intValue
            ),
            autoReload: autoReload,
            refreshInterval: refreshInterval.intValue,
            callbackId: command.callbackId
        )
    }

    @objc func showBannerAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('showBannerAd', [position]);
        bannerManager.showBannerAd(command)
    }

    @objc func hideBannerAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('hideBannerAd', []);
        bannerManager.hideBannerAd(command)
    }

    @objc func destroyBannerAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('destroyBannerAd', []);
        bannerManager.destroyBannerAd(command)
    }

    // MARK: - MREC Ad

    @objc func loadMRecAd(_ command: CDVInvokedUrlCommand) {
        // nativePromise('loadMRecAd', [autoReload ?? true, refreshInterval ?? 30]);
        let autoReload = command.arguments[0] as! Bool
        let refreshInterval = command.arguments[1] as! NSNumber
        mrecManager.loadAd(
            AdSize.mediumRectangle,
            autoReload: autoReload,
            refreshInterval: refreshInterval.intValue,
            callbackId: command.callbackId
        )
    }

    @objc func showMRecAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('showMRecAd', [position]);
        mrecManager.showBannerAd(command)
    }

    @objc func hideMRecAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('hideMRecAd', []);
        mrecManager.hideBannerAd(command)
    }

    @objc func destroyMRecAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('destroyMRecAd', []);
        mrecManager.destroyBannerAd(command)
    }

    // MARK: - App Open Ad

    @objc func loadAppOpenAd(_ command: CDVInvokedUrlCommand) {
        // nativePromise('loadAppOpenAd', [autoReload ?? false, autoShow ?? false]);
        let appOpen: CASAppOpen
        if let ad = appOpenManager.getAd() as? CASAppOpen {
            appOpen = ad
        } else {
            appOpen = CASAppOpen(casID: casId)
        }

        appOpenManager.loadAd(
            command.callbackId,
            autoload: command.arguments[0] as! Bool,
            ad: appOpen
        )

        appOpen.isAutoshowEnabled = command.arguments[1] as! Bool
    }

    @objc func isAppOpenAdLoaded(_ command: CDVInvokedUrlCommand) {
        // nativePromise('isAppOpenAdLoaded', []);
        appOpenManager.isAdLoaded(command.callbackId)
    }

    @objc func showAppOpenAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('showAppOpenAd', []);
        appOpenManager.showAd(
            command.callbackId,
            controller: viewController
        )
    }

    @objc func destroyAppOpenAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('destroyAppOpenAd', []);
        appOpenManager.destroyAd(command.callbackId)
    }

    // MARK: - Interstitial Ad

    @objc func loadInterstitialAd(_ command: CDVInvokedUrlCommand) {
        // nativePromise('loadInterstitialAd', [autoReload ?? false, autoShow ?? false, minInterval ?? 0]);
        let interstitial: CASInterstitial
        if let ad = interstitialManager.getAd() as? CASInterstitial {
            interstitial = ad
        } else {
            interstitial = CASInterstitial(casID: casId)
        }

        interstitialManager.loadAd(
            command.callbackId,
            autoload: command.arguments[0] as! Bool,
            ad: interstitial
        )
        interstitial.isAutoshowEnabled = command.arguments[1] as! Bool

        interstitial.minInterval = (command.arguments[2] as! NSNumber).intValue
    }

    @objc func isInterstitialAdLoaded(_ command: CDVInvokedUrlCommand) {
        // nativePromise('isInterstitialAdLoaded', []);
        interstitialManager.isAdLoaded(command.callbackId)
    }

    @objc func showInterstitialAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('showInterstitialAd', []);
        interstitialManager.showAd(
            command.callbackId,
            controller: viewController
        )
    }

    @objc func destroyInterstitialAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('destroyInterstitialAd', []);
        interstitialManager.destroyAd(command.callbackId)
    }

    // MARK: - Rewarded Ad

    @objc func loadRewardedAd(_ command: CDVInvokedUrlCommand) {
        // nativePromise('loadRewardedAd', [autoReload ?? false]);
        let rewarded: CASRewarded
        if let ad = rewardedManager.getAd() as? CASRewarded {
            rewarded = ad
        } else {
            rewarded = CASRewarded(casID: casId)
        }

        rewardedManager.loadAd(
            command.callbackId,
            autoload: command.arguments[0] as! Bool,
            ad: rewarded
        )
    }

    @objc func isRewardedAdLoaded(_ command: CDVInvokedUrlCommand) {
        // nativePromise('isRewardedAdLoaded', []);
        rewardedManager.isAdLoaded(command.callbackId)
    }

    @objc func showRewardedAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('showRewardedAd', []);
        rewardedManager.showAd(
            command.callbackId,
            controller: viewController
        )
    }

    @objc func destroyRewardedAd(_ command: CDVInvokedUrlCommand) {
        // nativeCall('destroyRewardedAd', []);
        rewardedManager.destroyAd(command.callbackId)
    }

    func getConsentFlowStatus(from model: ConsentFlowStatus) -> String {
        let message: String
        switch model {
        case .obtained:
            message = "Obtained"
        case .notRequired:
            message = "Not required"
        case .unavailable:
            message = "Unavailable"
        case .internalError:
            message = "Internal error"
        case .networkError:
            message = "Network error"
        case .viewControllerInvalid:
            message = "Invalid context"
        case .flowStillPresenting:
            message = "Still presenting"
        default:
            message = "Unknown"
        }
        return message
    }
}

// MARK: - Cordova Event Bridge

extension CASMobileAds {
    func sendOk(_ callbackId: String?) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK)
        commandDelegate?.send(result, callbackId: callbackId)
    }

    func sendOk(_ callbackId: String?, messageAs: String) {
        let result = CDVPluginResult(
            status: CDVCommandStatus_OK,
            messageAs: messageAs
        )
        commandDelegate?.send(result, callbackId: callbackId)
    }

    func sendOk(_ callbackId: String?, messageAs: Bool) {
        let result = CDVPluginResult(
            status: CDVCommandStatus_OK,
            messageAs: messageAs
        )
        commandDelegate?.send(result, callbackId: callbackId)
    }

    func sendOk(_ callbackId: String?, messageAs: [String: Any]) {
        let result = CDVPluginResult(
            status: CDVCommandStatus_OK,
            messageAs: messageAs
        )
        commandDelegate?.send(result, callbackId: callbackId)
    }

    func sendRejectError(_ callbackId: String?, format: String) {
        let body: [String: Any] = [
            "format": format,
            "code": 499,
            "message": "Load Promise interrupted by new load call",
        ]
        let result = CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: body
        )
        commandDelegate?.send(result, callbackId: callbackId)
    }

    func sendError(_ callbackId: String?, format: String, error: AdError) {
        let body: [String: Any] = [
            "format": format,
            "code": error.code.rawValue,
            "message": error.description,
        ]
        let result = CDVPluginResult(
            status: CDVCommandStatus_ERROR,
            messageAs: body
        )
        commandDelegate?.send(result, callbackId: callbackId)
    }

    func fireEvent(_ name: CASEvent, format: String) {
        fireDocumentEvent(name.rawValue, body: ["format": format])
    }

    func fireEvent(_ name: CASEvent, body: [String: Any]) {
        fireDocumentEvent(name.rawValue, body: body)
    }

    func fireImpressionEvent(format: String, contentInfo: AdContentInfo) {
        let revenuePrecision: String
        switch contentInfo.revenuePrecision {
        case .estimated:
            revenuePrecision = "estimated"
        case .precise:
            revenuePrecision = "precise"
        case .floor:
            revenuePrecision = "floor"
        default:
            revenuePrecision = "unknown"
        }

        var body: [String: Any] = [
            "format": format,
            "sourceUnitId": contentInfo.sourceUnitID,
            "sourceName": contentInfo.sourceName,
            "revenue": contentInfo.revenue,
            "revenuePrecision": revenuePrecision,
            "revenueTotal": contentInfo.revenueTotal,
            "impressionDepth": contentInfo.impressionDepth,
        ]
        body["creativeId"] = contentInfo.creativeID
        fireDocumentEvent(CASEvent.casai_ad_impressions.rawValue, body: body)
    }

    func fireErrorEvent(_ name: CASEvent, format: String, error: AdError) {
        let body: [String: Any] = [
            "format": format,
            "code": error.code.rawValue,
            "message": error.description,
        ]
        fireDocumentEvent(name.rawValue, body: body)
    }

    private func fireDocumentEvent(_ name: String, body: [String: Any]) {
        var jsonBody = "{}"
        if let data = try? JSONSerialization.data(
            withJSONObject: body,
            options: []
        ),
            let jsonString = String(data: data, encoding: .utf8)
        {
            jsonBody = jsonString
        }
        let js = "cordova.fireDocumentEvent('\(name)', \(jsonBody));"
        DispatchQueue.main.async {
            self.commandDelegate?.evalJs(js)
        }
    }
}
